### ISSUE TITLE
ZOOMA import times out intermittently

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#59]
I noticed that in deployment #51, but it will probably be true for any recent deployment.

Currently, ZOOMA query to update suggestions for all mappings (dummy + a selection of ClinVar which we import) takes just about 30 seconds to complete. Due to randomness in ZOOMA server response, sometimes it's less than 30s and everything's fine. However, when the time exceeds 30 seconds, the request times out:
![image](https://user-images.githubusercontent.com/10669118/86513958-3c5c8900-be06-11ea-942a-99beae137dd9.png)

```
[...]
2020-07-04T13:51:56.363771+00:00 app[web.1]: INFO:ZOOMA:Deleted mapping suggestions <QuerySet []>
2020-07-04T13:51:56.363905+00:00 app[web.1]: INFO:ZOOMA:Retrieving ZOOMA suggestions for trait: Lung adenocarcinoma
2020-07-04T13:51:56.400109+00:00 app[web.1]: [2020-07-04 13:51:56 +0000] [4] [CRITICAL] WORKER TIMEOUT (pid:14)
2020-07-04T13:51:56.401534+00:00 app[web.1]: [2020-07-04 13:51:56 +0000] [14] [INFO] Worker exiting (pid: 14)
```

To solve that, we will need to implement all long queries (ClinVar, ZOOMA etc.) to run in background, rather than waiting for them to complete inside a single request. Hopefully Heroku supports running something on the server in the background—otherwise we'll have to prioritise setting up another means of deployment.

**Comment by tskir** [comment]
@joj0s Could you please look into whether Heroku supports something like this? Ideally I'd like to know before our next catch-up call, so that we could discuss alternative deployment environments if necessary

**Comment by joj0s** [comment]
https://devcenter.heroku.com/categories/background-jobs-in-python
Looks like they are, I am mostly interested in using it with [APScheduler](https://devcenter.heroku.com/articles/clock-processes-python), although this is something I haven't done before and I am going to need to do some testing to confirm that

**Comment by joj0s** [comment]
I think I should actually prioritize this over the rest of the weekly goals

**Comment by tskir** [comment]
> https://devcenter.heroku.com/categories/background-jobs-in-python
> Looks like Heroku supports background jobs, I am mostly interested in using it with APScheduler, although this is something I haven't done before and I am going to need to do some testing to confirm that

OK, good. The only note I have is that whatever solution we choose it must not be limited to Heroku and we should be able to deploy it on other platform as well in the future (perhaps after configuration updates).

> I think I should actually prioritize this over the rest of the weekly goals
Agreed. Maybe not necessarily completely fixing this just yet, but it would be nice if you do some testing to see which solution will work.

**Comment by joj0s** [comment]
@tskir After some research, I think using [Celery](https://docs.celeryproject.org/en/latest/getting-started/first-steps-with-celery.html#tut-celery) with Redis as a message broker is a better choice for asynchronous tasks and scheduling. I am running some tests locally shortly, and in the next days I will open a PR which I will mark as WIP to test it on Heroku as well

**Comment by joj0s** [comment]
I can confirm local tests were successful in turning ClinVar and ZOOMA data retrieval to background tasks using Celery and Redis. Now there are some issues that remain with that:

- In order to test it, I used the free [Heroku Redis Addon](https://elements.heroku.com/addons/heroku-redis) on an empty app of mine. That being said, what we would want for now would be to create a free Redis add-on for each deployment and connect to that right? So I should research ways of automaticaly attaching the add-on upon deployment and possibly getting the URL from a Heroku env variable
- How should we handle the time between a task being initiated and it being finished? One approach that comes to mind is using progress bars in the datasources page, or maybe showing a notification when the app is done?

**Comment by tskir** [comment]
> I can confirm local tests were successful in turning ClinVar and ZOOMA data retrieval to background tasks using Celery and Redis.

Sounds good, let's use them.

> So I should research ways of automaticaly attaching the add-on upon deployment and possibly getting the URL from a Heroku env variable

Yes, let's try to do that. If it can be done automatically, then we can continue using Heroku for now. Otherwise, let me know and I'll prioritise the issue about procuring more mature and production-like environment for our deployments.

> How should we handle the time between a task being initiated and it being finished? One approach that comes to mind is using progress bars in the datasources page, or maybe showing a notification when the task is done?

Progress bars would be very nice, but I imagine they would be more difficult to implement. So for now we can just have a status indication. That is, when any import button is clicked, the process starts in the background and current status is displayed on the datasources page. This page can then auto-refresh every X seconds, with the status being updated.

**Comment by joj0s** [comment]
> Yes, let's try to do that. If it can be done automatically, then we can continue using Heroku for now. Otherwise, let me know and I'll prioritise the issue about procuring more mature and production-like environment for our deployments.

@tskir I can't seem to find a way to automatically attach the Heroku Redis add-on. Worst case scenario, we might need to do this manually but I don't think this will be much of a deal, as it would need to be done only once per deployment. The problem is that even though the add-on is free, there needs to be a verified credit/debit card attached to the owner's account in case the traffic exceeds the free tier limit (which it won't during development). So if it is doable for you let me know, otherwise maybe we could elevate my role in the Heroku pipeline and see if I can do it instead.

> Progress bars would be very nice, but I imagine they would be more difficult to implement. So for now we can just have a status indication. That is, when any import button is clicked, the process starts in the background and current status is displayed on the datasources page. This page can then auto-refresh every X seconds, with the status being updated.

Actually it might not be, thanks to this package https://github.com/czue/celery-progress. So I will go for that

**Comment by joj0s** [comment]
The progress bar also works. I am thinking that since I am already working on the datasources page, maybe we should change this week's goals and go with issue #32 as the next one.

**Comment by tskir** [comment]
> The problem is that even though the add-on is free, there needs to be a verified credit/debit card attached to the owner's account in case the traffic exceeds the free tier limit (which it won't during development).

I really wouldn't want to go this way. Cloud platforms are tricky in this regard. I have an example from one my previous places of work of someone's poorly written script getting stuck and costing upwards of $ 10,000 in cloud expenses for the company. Even though the possibility is remote, we still expose a public and writeable service, and I don't want to deal with _any_ finance-related concerns unless a proper accountability system is in place, which is clearly out of scope for our project.

Additionally, I feel that if we go beyond Heroku core functionality and start using its plugins, we will become more entangled with their environment. I also don't want for this to happen, since the app should be eventually deployable anywhere with proper set up, and the Heroku environment was only transitional to begin with.

Instead, here's what I think we should do:
* For an immediate solution, to avoid this being a blocker, let's decrease the number of traits being imported from ClinVar. We don't really need that many of them to develop nearly all functionality in this app. The requests currently take around or just over 30 seconds, so if you decrease the number of ClinVar-imported traits from 200 back to 100, it should continue working with no problems.
* We (or rather, I) should prioritise work on setting up a more proper deployment system to replace Heroku and migrate to that. I created a ticket, assigned to me, to start doing this: #61. I'll probably have some questions for you, which I'll ask in that ticket.

**Comment by tskir** [comment]
Oh yes, and regarding the progress bar — it's nice that its implementation is not as difficult.

> I am thinking that since I am already working on the datasources page, maybe we should change this week's goals and go with issue #32 as the next one. On the other hand, we could just keep the raw progress bar for now, and do that when we have OLS checks and configuration variables working.

Please do whichever seems more sensible, both options look good to me. Going with #32 is perhaps a good idea.

**Comment by joj0s** [comment]
> For an immediate solution, to avoid this being a blocker, let's decrease the number of traits being imported from ClinVar. We don't really need that many of them to develop nearly all functionality in this app. The requests currently take around or just over 30 seconds, so if you decrease the number of ClinVar-imported traits from 200 back to 100, it should continue working with no problems. We (or rather, I) should prioritise work on setting up a more proper deployment system to replace Heroku and migrate to that. I created a ticket, assigned to me, to start doing this: #61. I'll probably have some questions for you, which I'll ask in that ticket.

I think that those tasks being background tasks should be the case now that the code for them is mostly ready. So I would take responsibility using a Redis add-on attached to my account until we migrate environments, since the possibility of being charged is really unlikely, and besides, monthly costs are being constantly monitored I have a lot of free Heroku credits due to GitHub student program should something go wrong.

**Comment by joj0s** [comment]
@tskir Actually now that I think of it, I'm pretty sure it's impossible to exceed the limit, since the only connection being made to redis, is that of the celery instance and it is not done programmatically by me.

**Comment by tskir** [comment]
OK, well, if you insist :) To implement this, I would need to transfer the pipeline ownership to you. I've looked into this and it turns out it's not possible to transfer pipeline from a person to another person, [only from a person to a team](https://devcenter.heroku.com/articles/pipelines#pipelines-ownership-and-transfer). So if you'd like to go this way, then please [create a team](https://dashboard.heroku.com/teams/new), add a credit card, and then I would transfer ownership of the pipeline to that team. Although, for the record, I would still advise against doing that. _(Yes I might be paranoid, but then years in the industry tend to influence people in that way...)_

Regarding the proper environment: after I read and address your comments in #61, I'll write to the appropriate EBI team right away, and we should get a machine for deployment soon enough.

**Comment by joj0s** [comment]
Well, it turns out that teams are not allowed to own any free dynos, so that option is a no go as well. I think we should stick to inserting Redis URLs manually until the deployment situation is sorted out.

**Comment by tskir** [comment]
All right. I'll be requesting the instances today, and hopefully we'll be able to get access to them later this week.
